extras/reset_mbients.py

```diff
@@ -1,77 +1,309 @@
+        """
+        device = MetaWear(self.device_info.address)
+
+        success = False
+        self.logger.debug(self.format_message(f'Attempting Connection'))
+        for i in range(self.connect_attempts):
+            try:
+                if i > 0:  # Do not immediately try to reconnect if it just failed.
+                    sleep(3)
+
+                device.connect()
+                success = True
+                break
+            except Exception as e:
+                self.logger.debug(self.format_message(f'Failed to connect on attempt {i + 1}: {e}'))
 
-if __name__ == "__main__":
-    macs = {
-        "Mbient_LH_2": "E8:95:D6:F7:39:D2",
-        "Mbient_RH_2": "FE:07:3E:37:F5:9C",
-        "Mbient_RF_2": "E5:F6:FB:6D:11:8A",
-        "Mbient_LF_2": "DA:B0:96:E4:7F:A3",
-        "Mbient_BK_1": "D7:B0:7E:C2:A1:23",
-        }
-    
-    print('resetting mbients (will take ~ 1 min)...')
-    scann_BLE(2)
-    for k, v in macs.items():
-        success = reset_dev(v)
         if not success:
-            print(f"Failed to connect {k} {v}")
-        else:
-            print(f"Success in resetting {k} {v}")
+            raise Exception(self.format_message(f'Unable to connect!'))
+
+        device.on_disconnect = lambda status: self.on_disconnect(status)
+        return device
+
+    def on_disconnect(self, status) -> None:
+        self.logger.debug(self.format_message(f'Disconnected with status {status}.'))
+        self.disconnect_event.set()
+
+    @staticmethod
+    def reset_device(device: MetaWear) -> None:
+        """
+        Reset the device. See https://mbientlab.com/tutorials/PyLinux.html#reset
+
+        :param device: The connected device object to reset
+        """
+        board = device.board
+        libmetawear.mbl_mw_logging_stop(board)
+        libmetawear.mbl_mw_logging_flush_page(board)
+        libmetawear.mbl_mw_logging_clear_entries(board)
+        libmetawear.mbl_mw_event_remove_all(board)
+        libmetawear.mbl_mw_macro_erase_all(board)
+        libmetawear.mbl_mw_debug_reset_after_gc(board)
+        libmetawear.mbl_mw_debug_disconnect(board)
+
+
+def reset_devices(args: argparse.Namespace, devices: ADDRESS_MAP) -> None:
+    reset_threads = [
+        ResetDeviceThread(
+            device_info=device,
+            connect_attempts=args.n_connect_attempts,
+            reset_timeout=args.reset_timeout,
+        ) for device in devices
+    ]
 
-    sleep(60)
+    logger = logging.getLogger('default')
+    logger.info(f'Reset in progress...')
+    t0 = time()
+    for t in reset_threads:
+        t.start()
+    for t in reset_threads:
+        t.join()
+    logger.info(f'Device reset {time() - t0:0.1f} sec.')
+
+    success = [t.device_info for t in reset_threads if t.success]
+    failure = [t.device_info for t in reset_threads if not t.success]
+    logger.info(f'Successfully reset {len(success)} devices:')
+    print_device_info(success, print_fn=logger.info)
```

Comment:
maybe return a list of the failed devices? At some point, I imagine this may be (a) part of the UI, or (b) part of retry loop where only the failed devices get reset.

Reply:
Agreed, I'll return the list and for now do nothing with it. I think the reset script mostly failed when attempting to connect, which is why I currently don't have the retries for the reset.

Reply:
Decided to return the list of success and failures, and handled the log output in main.